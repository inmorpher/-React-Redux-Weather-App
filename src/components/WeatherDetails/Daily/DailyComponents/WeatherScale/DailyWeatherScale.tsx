import {
	forwardRef,
	ReactNode,
	SVGLineElementAttributes,
	SVGProps,
	SVGTextElementAttributes,
} from 'react';

/**
 * A component that provides a container for daily weather scale elements.
 *
 * @param children - The child elements to be rendered within the container.
 * @returns A div element with the specified children, styled to center the content.
 */
const DailyWeatherScale = ({ children }: { children: ReactNode }) => {
	return <div className='mt-2 flex justify-center'>{children}</div>;
};

export const SVGWrapper = (props: SVGProps<SVGSVGElement>) => <svg {...props} />;

export const Line = (props: SVGLineElementAttributes<SVGLineElement>) => (
	<line {...props} data-testid='weather-curve-line' />
);

export const Text = (props: SVGTextElementAttributes<SVGTextElement>) => (
	<text {...props} data-testid='weather-curve-text' />
);

export const Curve = forwardRef<SVGPathElement, SVGProps<SVGPathElement>>((props, ref) => (
	<path {...props} ref={ref} />
));

DailyWeatherScale.SVGWrapper = SVGWrapper;
DailyWeatherScale.Line = Line;
DailyWeatherScale.Text = Text;
DailyWeatherScale.Curve = Curve;

export default DailyWeatherScale;
