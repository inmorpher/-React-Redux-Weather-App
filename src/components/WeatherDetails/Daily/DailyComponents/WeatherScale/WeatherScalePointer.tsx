import { RefObject } from 'react';
import { z } from 'zod';
import { MetricReturnType } from '../../../../../utils/services/converter/metric.converter';
import usePopupWeatherScale from './usePopupWeatherScale';

const WeatherScalePropsSchema = z.object({
	curve: z.custom<RefObject<SVGPathElement>>(
		(val) => val !== null && typeof val === 'object' && 'current' in val
	),
	data: z.array(z.custom<MetricReturnType>()),
});

type WeatherScaleProps = z.infer<typeof WeatherScalePropsSchema>;

const WeatherScalePointer = ({ curve, data }: WeatherScaleProps) => {
	const propsParseResult = WeatherScalePropsSchema.safeParse({ curve, data });
	const hookResult = usePopupWeatherScale(curve, data);
	const { state, pointerOnCurveMoveHandler, pointerOnCurveIn, pointerOnCurveOut } =
		hookResult ?? {};

	return (
		<svg>
			<g data-popup='popup-pointer'>
				{state?.pointerVisibility && propsParseResult.success && (
					<g>
						<text
							x={state.pointer?.x ?? 0}
							y={15}
							fill='#fff'
							fontSize='.6rem'
							fontWeight='700'
							dominantBaseline='bottom'
							textAnchor='middle'
						>
							{state.pointerTempriture ?? ''}
							<tspan x={state.pointer?.x ?? 0} dy={10} fontWeight={300} fontSize={'.4rem'}>
								{state.pointerTime ?? ''}
							</tspan>
						</text>
						<circle cx={state.pointer?.x ?? 0} cy={state.pointer?.y ?? 0} r='5' fill='#fff' />
						<line
							x1={state.pointer?.x ?? 0}
							y1={30}
							x2={state.pointer?.x ?? 0}
							y2={140}
							stroke='#fff'
							strokeWidth={0.5}
							strokeLinecap='round'
						/>
					</g>
				)}
				<rect
					x={0}
					y={0}
					height={160}
					width={320}
					fill='none'
					stroke='none'
					onMouseMove={pointerOnCurveMoveHandler ?? undefined}
					onTouchMove={pointerOnCurveMoveHandler ?? undefined}
					onMouseEnter={pointerOnCurveIn ?? undefined}
					onMouseLeave={pointerOnCurveOut ?? undefined}
					onTouchStart={pointerOnCurveIn ?? undefined}
					onTouchEnd={pointerOnCurveOut ?? undefined}
					pointerEvents={'all'}
					role='presentation'
				/>
			</g>
		</svg>
	);
};

export default WeatherScalePointer;
