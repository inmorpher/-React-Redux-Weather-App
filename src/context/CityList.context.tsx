import { createContext, useContext, useEffect, useState } from 'react';
import { z } from 'zod';
export interface ICityList {
	city: string;
	country: string;
	state?: string;
	lat: string | number;
	lon: string | number;
}

export interface ICityListData {
	version: number;
	list: ICityList[];
}

interface CityListContextType extends ICityListData {
	showDeleteBtn: boolean;
	addCity: (city: ICityList) => void;
	toggleDeleteBtn: () => void;
	deleteCity: (city: ICityList) => void;
}

const CITY_LIST_VERSION = 1 as const;

const CityListContext = createContext<CityListContextType | undefined>(undefined);

/**
 * Provides the city list context to its children.
 *
 * @param children - The child components that will have access to the city list context.
 * @returns A context provider component that supplies the city list data and functions to its children.
 */
export const CityListProvider = ({ children }: { children: React.ReactNode }) => {
	const CitySchema = z.object({
		city: z.string().min(1),
		country: z.string().min(1),
		state: z.string().min(1).optional(),
		lat: z
			.union([z.number(), z.string()])
			.refine((val) => val !== '', { message: 'Latitude cannot be empty' }),
		lon: z
			.union([z.number(), z.string()])
			.refine((val) => val !== '', { message: 'Longitude cannot be empty' }),
	});
	const [cityList, setCityList] = useState<ICityListData>(() => {
		const savedCityList = localStorage.getItem('cityList');
		if (savedCityList) {
			const parsedCityList = JSON.parse(savedCityList);
			if (parsedCityList.version === CITY_LIST_VERSION) {
				return parsedCityList;
			}
		}

		return {
			version: CITY_LIST_VERSION,
			list: [],
		};
	});

	const [showDeleteBtn, setShowDeleteBtn] = useState<boolean>(false);

	/**
	 * Toggles the visibility of the delete button.
	 */
	const toggleDeleteBtn = () => {
		setShowDeleteBtn((prevShowDeleteBtn) => !prevShowDeleteBtn);
	};

	/**
	 * Adds a new city to the city list if it does not already exist.
	 *
	 * @param city - The city to be added to the list.
	 */
	// const addCity = (city: ICityList) => {
	// 	const cityExists = cityList.list.some(
	// 		(existingCity) =>
	// 			existingCity.city === city.city &&
	// 			existingCity.country === city.country &&
	// 			existingCity.state === city.state &&
	// 			existingCity.lat === city.lat &&
	// 			existingCity.lon === city.lon
	// 	);

	// 	if (!cityExists) {
	// 		setCityList((prevCityList) => ({
	// 			...prevCityList,
	// 			list: [...prevCityList.list, city],
	// 		}));
	// 	}
	// };
	const addCity = async (city: ICityList) => {
		try {
			const validatedCity = CitySchema.parse(city);
			setCityList((prevCityList) => {
				const cityExists = prevCityList.list.some(
					(existingCity) =>
						existingCity.city === validatedCity.city &&
						existingCity.country === validatedCity.country &&
						existingCity.state === validatedCity.state &&
						existingCity.lat === validatedCity.lat &&
						existingCity.lon === validatedCity.lon
				);

				if (!cityExists) {
					return {
						...prevCityList,
						list: [...prevCityList.list, validatedCity],
					};
				}
				return prevCityList;
			});
		} catch (error) {
			console.error('Invalid city data:', error);
		}
	};
	/**
	 * Deletes a city from the city list.
	 *
	 * @param city - The city to be deleted from the list.
	 */
	const deleteCity = (city: ICityList) => {
		setCityList((prevCityList) => ({
			...prevCityList,
			list: prevCityList.list.filter(
				(existingCity) =>
					existingCity.city !== city.city ||
					existingCity.country !== city.country ||
					existingCity.state !== city.state ||
					existingCity.lat !== city.lat ||
					existingCity.lon !== city.lon
			),
		}));
	};

	useEffect(() => {
		localStorage.setItem('cityList', JSON.stringify(cityList));
	}, [cityList.list.length]);

	return (
		<CityListContext.Provider
			value={{ ...cityList, addCity, showDeleteBtn, toggleDeleteBtn, deleteCity }}
		>
			{children}
		</CityListContext.Provider>
	);
};

export const useCityList = () => {
	const context = useContext(CityListContext);
	if (!context) {
		throw new Error('useCityList must be used within a CityListProvider');
	}
	return context;
};
