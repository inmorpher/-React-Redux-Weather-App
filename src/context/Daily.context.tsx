import React, {
	ReactNode,
	createContext,
	startTransition,
	useCallback,
	useRef,
	useState,
} from 'react';
import { useWindowResize } from '../hooks/useWindowResize';

export interface DailyContextProps {
	children: ReactNode;
}

export interface DailyState {
	isOpen: boolean;
	item: number;
}

export interface DailyContextType {
	dailyState: DailyState;
	isDailyOpen: boolean;
	activeDay: number;
	isActiveListener: boolean;
	containerRef: React.RefObject<HTMLDivElement>;
	dailyListRef: React.RefObject<HTMLDivElement>;
	dailyDetailsRef: React.RefObject<HTMLDivElement>;
	scrollToggler: () => void;
	hideDetails: () => void;
	showDetails: (item: number) => void;
	setIsActiveListener: (value: boolean) => void;
	onOpenPopup: (item: number) => void;
	onPressKeys: (event: React.KeyboardEvent<HTMLLIElement>, index: number) => void;
	onCloseDetails: () => void;
}

export const DailyContext = createContext<DailyContextType | undefined>(undefined);

export const DailyProvider: React.FC<DailyContextProps> = ({ children }) => {
	const [dailyState, setDailyState] = useState<DailyState>({
		isOpen: false,
		item: 0,
	});
	const [isDailyOpen] = useState(false);
	const [activeDay] = useState<number>(0);

	const [isActiveListener, setIsActiveListener] = useState(false);

	const containerRef = useRef<HTMLDivElement>(null);
	const dailyListRef = useRef<HTMLDivElement>(null);
	const dailyDetailsRef = useRef<HTMLDivElement>(null);
	/**
	 * Opens the popup with the given item and position.
	 * @param item the index of the item to display in the popup
	 * @param position the position of the popup, or 'mobile' to use the mobile layout
	 */

	const showDetails = useCallback((item: number) => {
		startTransition(() => {
			setDailyState({
				isOpen: true,
				item,
			});
		});
	}, []);

	const hideDetails = () => {
		startTransition(() => {
			setDailyState(() => ({
				item: 0,
				isOpen: false,
			}));
		});
	};

	const scrollToggler = useCallback(() => {
		if (containerRef.current && dailyListRef.current && dailyDetailsRef.current) {
			const containerLeft = containerRef.current.offsetLeft;
			const dailyDetailsLeft = dailyDetailsRef.current.offsetLeft;
			containerRef.current.scrollTo({
				left: dailyDetailsLeft - containerLeft,
				behavior: 'smooth',
			});
		}
	}, [containerRef, dailyDetailsRef, dailyListRef]);

	/**
	 * Handles the opening of the popup for daily details.
	 * It triggers the detail view to show and scrolls the container accordingly.
	 *
	 * @param {number} index - The index of the day for which details should be shown.
	 */
	const onOpenPopup = useCallback(
		(index: number) => {
			showDetails(index);
			scrollToggler();
		},
		[showDetails, scrollToggler]
	);

	useWindowResize(() => {
		if (dailyState.isOpen) {
			scrollToggler();
		}
	}, dailyState.isOpen);

	/**
	 * Closes the daily details view and scrolls the container back to the start.
	 */
	const onCloseDetails = useCallback(() => {
		if (containerRef.current) {
			hideDetails();

			containerRef.current?.scrollTo({
				left: 0,
				behavior: 'smooth',
			});
		}
	}, [containerRef, hideDetails]);
	/**
	 * Handles keyboard events for accessibility.
	 * It allows opening and closing of daily details using the Enter and Escape keys, respectively.
	 *
	 * @param {React.KeyboardEvent<HTMLLIElement>} event - The keyboard event.
	 * @param {number} index - The index of the day for which the keyboard event is triggered.
	 */
	const onPressKeys = useCallback(
		(event: React.KeyboardEvent<HTMLLIElement>, index: number) => {
			if (event.key === 'Enter' && !dailyState.isOpen) {
				onOpenPopup(index);
			} else if (event.key === 'Escape' && dailyState.isOpen) {
				onCloseDetails();
			}
		},
		[dailyState.isOpen, onCloseDetails, onOpenPopup]
	);
	return (
		<DailyContext.Provider
			value={{
				isDailyOpen,
				activeDay,
				dailyState,
				showDetails,
				hideDetails,
				containerRef,
				dailyDetailsRef,
				dailyListRef,
				isActiveListener,
				setIsActiveListener,
				onOpenPopup,
				onPressKeys,
				onCloseDetails,
				scrollToggler,
			}}
		>
			{children}
		</DailyContext.Provider>
	);
};
