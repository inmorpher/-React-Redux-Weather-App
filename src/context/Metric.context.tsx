import { createContext, useContext, useEffect, useState } from 'react';

export type MetricType = 'metric' | 'imperial';

export interface IUserMetric {
	version: number;
	metricType: MetricType;
}

interface MetricContextType extends IUserMetric {
	toggleMetric: () => void;
}

const METRIC_VERSION = 1 as const;

const MetricContext = createContext<MetricContextType | undefined>(undefined);

/**
 * MetricProvider is a context provider component that manages the user's metric preferences.
 * It provides the current metric type and a function to toggle between metric and imperial units.
 *
 * @param children - The child components that will have access to the metric context.
 * @returns A context provider component that supplies the current metric state and a toggle function.
 */
export const MetricProvider = ({ children }: { children: React.ReactNode }) => {
	const [metric, setMetric] = useState<IUserMetric>(() => {
		const savedMetric = localStorage.getItem('metricData');
		if (savedMetric) {
			const parsedMetric = JSON.parse(savedMetric);
			if (parsedMetric.version === METRIC_VERSION) {
				return parsedMetric;
			}
		}
		return { version: METRIC_VERSION, metricType: 'metric' };
	});

	useEffect(() => {
		localStorage.setItem('metricData', JSON.stringify(metric));
	}, [metric]);

	const toggleMetric = () => {
		setMetric((prevMetric) => ({
			...prevMetric,
			metricType: prevMetric.metricType === 'metric' ? 'imperial' : 'metric',
		}));
	};

	return (
		<MetricContext.Provider value={{ ...metric, toggleMetric }}>{children}</MetricContext.Provider>
	);
};

/**
 * useMetric is a custom hook that provides access to the metric context.
 * It throws an error if used outside of a MetricProvider.
 *
 * @returns The current metric state and a function to toggle the metric type.
 */
export const useMetric = (): MetricContextType => {
	const context = useContext(MetricContext);
	if (!context) {
		throw new Error('useMetric must be used within a MetricProvider');
	}
	return context;
};
