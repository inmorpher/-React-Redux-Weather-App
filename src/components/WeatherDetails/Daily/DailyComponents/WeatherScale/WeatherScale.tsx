import React, { useRef } from 'react';
import { z } from 'zod';
import { useGetDailyScale } from '../../../../../context/WeatherData.context';
import { useSelectedDayIndex } from '../../../../../hooks/useDaily';
import DailyWeatherScale from './DailyWeatherScale';
import WeatherScalePointer from './WeatherScalePointer';

// Define Zod schemas
const TemperatureCurveSchema = z.object({
	mainCurve: z.string(),
});

const TemperatureScaleItemSchema = z.object({
	pY: z.number(),
	value: z.number(),
	units: z.string(),
});

const ScaleDescriptionSchema = z.object({
	x: z.number(),
	y: z.number(),
	value: z.string(),
});

const ExpandedTemperatureDataSchema = z.object({
	value: z.number(),
	units: z.string(),
});

const DailyScaleSchema = z.object({
	temperatureCurve: TemperatureCurveSchema,
	temperatureScale: z.array(TemperatureScaleItemSchema),
	scaleDescriptions: z.array(ScaleDescriptionSchema),
	expandedTemperatureData: z.array(ExpandedTemperatureDataSchema),
});

const WeatherScale = () => {
	const selectedDayIndex = useSelectedDayIndex();
	const rawDailyScale = useGetDailyScale(selectedDayIndex) || {};

	const parsedForecast = DailyScaleSchema.safeParse(rawDailyScale);
	const validatedData = parsedForecast.success ? parsedForecast.data : null;

	const curveRef = useRef<SVGPathElement>(null);
	const isScale = parsedForecast.success && validatedData;
	return (
		<DailyWeatherScale>
			<DailyWeatherScale.SVGWrapper
				viewBox='0 0 320 160'
				className='bg-weather-gradient w-[90%] rounded-lg'
			>
				{isScale && (
					<>
						{validatedData.temperatureScale.map((item, index) => (
							<React.Fragment key={'PopupLayout' + index}>
								<DailyWeatherScale.Line
									key={'PopupScaleLine' + index}
									x1={15}
									y1={item.pY - 2.5}
									x2={300}
									y2={item.pY - 2.5}
									stroke='white'
									strokeDasharray={2}
									strokeDashoffset={10}
									strokeWidth={0.1}
								/>
								<DailyWeatherScale.Text
									key={'popupDesc' + index}
									x={10}
									y={item.pY}
									fill='#fff'
									fontSize='.4rem'
									fontWeight='300'
									dominantBaseline='bottom'
									textAnchor='middle'
								>
									{item.value}
									{item.units}
								</DailyWeatherScale.Text>
								<DailyWeatherScale.Text
									key={'PopupText' + index}
									x={validatedData.scaleDescriptions[index]?.x ?? 0}
									y={validatedData.scaleDescriptions[index]?.y ?? 0}
									fill='#fff'
									fontSize='.4rem'
									fontWeight='300'
									dominantBaseline='top'
									textAnchor='middle'
								>
									{validatedData.scaleDescriptions[index]?.value ?? 'N/A'}
								</DailyWeatherScale.Text>
								<DailyWeatherScale.Line
									key={'popupHorizontalLine' + index}
									x1={validatedData.scaleDescriptions[index]?.x ?? 0}
									y1={30}
									x2={validatedData.scaleDescriptions[index]?.x ?? 0}
									y2={140}
									stroke='white'
									strokeWidth={0.1}
								/>
							</React.Fragment>
						))}
						<DailyWeatherScale.Curve
							ref={curveRef}
							d={validatedData.temperatureCurve?.mainCurve ?? ''}
							stroke='#fff'
							fill='none'
							strokeWidth={2.5}
							strokeLinecap='round'
							strokeLinejoin='round'
							data-testid='main-curve'
						/>
						<WeatherScalePointer curve={curveRef} data={validatedData.expandedTemperatureData} />
					</>
				)}
			</DailyWeatherScale.SVGWrapper>
		</DailyWeatherScale>
	);
};

export default WeatherScale;
