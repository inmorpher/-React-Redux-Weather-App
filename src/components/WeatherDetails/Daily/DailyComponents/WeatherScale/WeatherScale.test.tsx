import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { MockedFunction, vi } from 'vitest';
import { useGetDailyScale } from '../../../../../context/WeatherData.context';
import { useSelectedDayIndex } from '../../../../../hooks/useDaily';
import WeatherScale from './WeatherScale';

vi.mock('../../../../../context/WeatherData.context', () => ({
	useGetDailyScale: vi.fn(),
}));

vi.mock('../../../../../hooks/useDaily', () => ({
	useSelectedDayIndex: vi.fn(),
}));

describe('WeatherScale', () => {
	// test('should handle the case when the selectedDayIndex is out of range', () => {
	// 	const mockData = {
	// 		temperatureCurve: { mainCurve: '' },
	// 		temperatureScale: [],
	// 		scaleDescriptions: [],
	// 		expandedTemperatureData: [],
	// 	} as unknown as ReturnType<typeof useGetDailyScale>;

	// 	(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(mockData);
	// 	(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(100);

	// 	const { container } = render(<WeatherScale />);

	// 	expect(container).toMatchSnapshot();
	// });

	test('should handle the case when useGetDailyScale returns undefined or null', () => {
		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			null as unknown as ReturnType<typeof useGetDailyScale>
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { container } = render(<WeatherScale />);
		expect(container.firstChild).toMatchSnapshot();

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			undefined as unknown as ReturnType<typeof useGetDailyScale>
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { container: nullContainer } = render(<WeatherScale />);
		expect(nullContainer.firstChild).toMatchSnapshot();
	});

	test('should render the correct number of scale lines and descriptions based on the temperatureScale array length', () => {
		const mockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [
				{ pY: 150, value: 0, units: '°C' },
				{ pY: 120, value: 10, units: '°C' },
				{ pY: 90, value: 20, units: '°C' },
				{ pY: 60, value: 30, units: '°C' },
			],
			scaleDescriptions: [
				{ x: 20, y: 140, value: 'Freezing' },
				{ x: 50, y: 110, value: 'Cool' },
				{ x: 80, y: 80, value: 'Warm' },
				{ x: 110, y: 50, value: 'Hot' },
			],
			expandedTemperatureData: [
				{ value: 0, units: '°C' },
				{ value: 10, units: '°C' },
				{ value: 20, units: '°C' },
				{ value: 30, units: '°C' },
			],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(mockData);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { getAllByTestId } = render(<WeatherScale />);

		const scaleLines = getAllByTestId('weather-curve-line');
		const scaleDescriptions = getAllByTestId('weather-curve-text');

		const expectedLength =
			(mockData?.temperatureScale?.length ?? 0) + (mockData?.scaleDescriptions?.length ?? 0);
		expect(scaleLines).toHaveLength(expectedLength);
		expect(scaleDescriptions).toHaveLength(expectedLength);
	});

	test('should handle the case when temperatureCurve.mainCurve is an empty string or invalid path data', () => {
		const invalidMockData = {
			temperatureCurve: { mainCurve: '' },
			temperatureScale: [],
			scaleDescriptions: [],
			expandedTemperatureData: [],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(invalidMockData);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { getByTestId, rerender } = render(<WeatherScale />);
		const curveElement = getByTestId('main-curve');

		expect(curveElement).toHaveAttribute('d', '');

		const invalidPathMockData = {
			temperatureCurve: { mainCurve: 'INVALID_PATH_DATA' },
			temperatureScale: [],
			scaleDescriptions: [],
			expandedTemperatureData: [],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			invalidPathMockData
		);

		rerender(<WeatherScale />);
		const invalidCurveElement = getByTestId('main-curve');

		expect(invalidCurveElement).toBeInTheDocument();
	});

	test('should handle the case when temperatureScale, scaleDescriptions, or expandedTemperatureData arrays are empty', () => {
		const emptyArraysMockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [],
			scaleDescriptions: [],
			expandedTemperatureData: [],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			emptyArraysMockData
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { queryByTestId } = render(<WeatherScale />);

		const scaleLines = queryByTestId('weather-curve-line');
		const scaleDescriptions = queryByTestId('weather-curve-text');

		expect(scaleLines).toBeNull();
		expect(scaleDescriptions).toBeNull();

		const mainCurveElement = queryByTestId('main-curve');
		expect(mainCurveElement).toBeInTheDocument();
	});

	test('should handle the case when temperatureScale, scaleDescriptions, or expandedTemperatureData arrays are empty', () => {
		const emptyArraysMockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [],
			scaleDescriptions: [],
			expandedTemperatureData: [],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			emptyArraysMockData
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { queryByTestId } = render(<WeatherScale />);

		const scaleLines = queryByTestId('weather-curve-line');
		const scaleDescriptions = queryByTestId('weather-curve-text');

		expect(scaleLines).toBeNull();
		expect(scaleDescriptions).toBeNull();

		const mainCurveElement = queryByTestId('main-curve');
		expect(mainCurveElement).toBeInTheDocument();
	});

	test('should handle the case when the units in temperatureScale or expandedTemperatureData are missing or invalid', () => {
		const missingUnitsMockData = null;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			missingUnitsMockData as unknown as ReturnType<typeof useGetDailyScale>
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { container } = render(<WeatherScale />);

		const svgElement = container.querySelector('svg');
		expect(svgElement).toBeInTheDocument();
		expect(svgElement?.children.length).toBe(0);

		const scaleLines = container.querySelectorAll('[data-testid="weather-curve-line"]');
		const scaleDescriptions = container.querySelectorAll('[data-testid="weather-curve-text"]');

		expect(scaleLines.length).toBe(0);
		expect(scaleDescriptions.length).toBe(0);
	});
	test('should render the main curve correctly', () => {
		const mockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [],
			scaleDescriptions: [],
			expandedTemperatureData: [],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(mockData);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { getByTestId } = render(<WeatherScale />);

		const curveElement = getByTestId('main-curve');
		expect(curveElement).toHaveAttribute('d', mockData?.temperatureCurve.mainCurve);
	});
	test('should validate the data with Zod and render the component correctly', () => {
		const mockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [
				{ pY: 150, value: 0, units: '°C' },
				{ pY: 120, value: 10, units: '°C' },
				{ pY: 90, value: 20, units: '°C' },
				{ pY: 60, value: 30, units: '°C' },
			],
			scaleDescriptions: [
				{ x: 20, y: 140, value: 'Freezing' },
				{ x: 50, y: 110, value: 'Cool' },
				{ x: 80, y: 80, value: 'Warm' },
				{ x: 110, y: 50, value: 'Hot' },
			],
			expandedTemperatureData: [
				{ value: 0, units: '°C' },
				{ value: 10, units: '°C' },
				{ value: 20, units: '°C' },
				{ value: 30, units: '°C' },
			],
		};

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(
			mockData as unknown as ReturnType<typeof useGetDailyScale>
		);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { getByTestId, getAllByTestId } = render(<WeatherScale />);

		const curveElement = getByTestId('main-curve');
		expect(curveElement).toHaveAttribute('d', mockData.temperatureCurve.mainCurve);

		const scaleLines = getAllByTestId('weather-curve-line');
		const scaleDescriptions = getAllByTestId('weather-curve-text');

		expect(scaleLines).toHaveLength(
			mockData.temperatureScale.length + mockData.scaleDescriptions.length
		);
		expect(scaleDescriptions).toHaveLength(
			mockData.temperatureScale.length + mockData.scaleDescriptions.length
		);
	});

	test('should render the correct number of scale lines and descriptions based on the temperatureScale array length', () => {
		const mockData = {
			temperatureCurve: { mainCurve: 'M10 150 C 30 120, 70 90, 100 50' },
			temperatureScale: [
				{ pY: 150, value: 0, units: '°C' },
				{ pY: 120, value: 10, units: '°C' },
				{ pY: 90, value: 20, units: '°C' },
				{ pY: 60, value: 30, units: '°C' },
			],
			scaleDescriptions: [
				{ x: 20, y: 140, value: 'Freezing' },
				{ x: 50, y: 110, value: 'Cool' },
				{ x: 80, y: 80, value: 'Warm' },
				{ x: 110, y: 50, value: 'Hot' },
			],
			expandedTemperatureData: [
				{ value: 0, units: '°C' },
				{ value: 10, units: '°C' },
				{ value: 20, units: '°C' },
				{ value: 30, units: '°C' },
			],
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(mockData);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { getAllByTestId } = render(<WeatherScale />);

		const scaleLines = getAllByTestId('weather-curve-line');
		const scaleDescriptions = getAllByTestId('weather-curve-text');

		const expectedLength =
			(mockData?.temperatureScale?.length ?? 0) + (mockData?.scaleDescriptions?.length ?? 0);
		expect(scaleLines).toHaveLength(expectedLength);
		expect(scaleDescriptions).toHaveLength(expectedLength);
	});

	test('should handle the case when Zod validation fails', () => {
		// Mock data that does not conform to the Zod schema
		const invalidMockData = {
			temperatureCurve: { mainCurve: 123 }, // Invalid type
			temperatureScale: [{ pY: 'invalid', value: 'invalid', units: 123 }], // Invalid types
			scaleDescriptions: [{ x: 'invalid', y: 'invalid', value: 123 }], // Invalid types
			expandedTemperatureData: [{ value: 'invalid', units: 123 }], // Invalid types
		} as unknown as ReturnType<typeof useGetDailyScale>;

		(useGetDailyScale as MockedFunction<typeof useGetDailyScale>).mockReturnValue(invalidMockData);
		(useSelectedDayIndex as MockedFunction<typeof useSelectedDayIndex>).mockReturnValue(0);

		const { container } = render(<WeatherScale />);
		const svgElement = container.querySelector('svg');
		expect(svgElement).toBeInTheDocument();
		expect(svgElement?.children.length).toBe(0);
	});
});
