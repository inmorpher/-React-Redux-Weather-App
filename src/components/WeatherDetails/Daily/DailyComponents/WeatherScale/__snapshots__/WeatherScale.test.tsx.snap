// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WeatherScale > should handle the case when the selectedDayIndex is out of range 1`] = `
<div>
  <div
    class="mt-2 flex justify-center"
  >
    <svg
      class="bg-weather-gradient w-[90%] rounded-lg"
      viewBox="0 0 320 160"
    >
      <path
        d=""
        data-testid="main-curve"
        fill="none"
        stroke="#fff"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2.5"
      />
      <g
        data-popup="popup-pointer"
      >
        <rect
          fill="none"
          height="160"
          pointer-events="all"
          stroke="none"
          width="320"
          x="0"
          y="0"
        />
      </g>
    </svg>
  </div>
</div>
`;

exports[`WeatherScale > should handle the case when useGetDailyScale returns undefined or null 1`] = `
<div
  class="mt-2 flex justify-center"
>
  <svg
    class="bg-weather-gradient w-[90%] rounded-lg"
    viewBox="0 0 320 160"
  />
</div>
`;

exports[`WeatherScale > should handle the case when useGetDailyScale returns undefined or null 2`] = `
<div
  class="mt-2 flex justify-center"
>
  <svg
    class="bg-weather-gradient w-[90%] rounded-lg"
    viewBox="0 0 320 160"
  />
</div>
`;
